File: C-Abi-Bright-Aot/Tests/CAbiBridgeAot.Tests/TestCrypto/TestCryptoChaCha20Poly1305/CryptoChaCha20Poly1305BytesTest.cs

 


using System.Text; 
using System.Diagnostics;
using System.Security.Cryptography;

namespace michele.natale.Tests;

using Pointers; 
using CAbiBridge;
using static CryptoTestUtils;

partial class CryptoChaCha20Poly1305Test
{

  #region Native
  public static void TestChaCha20Poly1305Bytes(int rounds)
  {
    Console.Write($"{nameof(TestChaCha20Poly1305Bytes)}: ");

    var rand = Random.Shared;
    var sw = Stopwatch.StartNew();
    for (int i = 0; i < rounds; i++)
    {
      var size = rand.Next(
        NetServices.MIN_PW_SIZE, 17);

      var pw = RngBytes(size);

      size = rand.Next(
        NetServices.MIN_SALT_SIZE,
        NetServices.MAX_SALT_SIZE);

      var salt = RngBytes(size);

      using var key = new UsIPtr<byte>(
        Rfc2898DeriveBytes.Pbkdf2(
        password: pw, salt: salt,
        iterations: NetServices.MIN_ITERATION,
        hashAlgorithm: HashAlgorithmName.SHA512,
        outputLength: NetServices.CHACHA_POLY_MAX_KEY_SIZE));

      var tag = new byte[NetServices.CHACHA_POLY_TAG_SIZE];
      var nonce = RngBytes(NetServices.CHACHA_POLY_NONCE_SIZE);

      size = rand.Next(
        NetServices.CHACHA_POLY_MIN_PLAIN_SIZE,
        NetServices.CHACHA_POLY_MAX_PLAIN_SIZE);

      var plain = RngBytes(size);

      var associat = Encoding.UTF8.GetBytes("© Michele Natale 2021");

      var err = Native.ChaCha20Poly1305EncryptAot(
      plain, plain.Length, key.Ptr, key.Length,
      associat, associat.Length,
      out IntPtr cipher_ptr, out int cipher_length);

      AssertError(err);
      var cipher = ToBytes(cipher_ptr, cipher_length);
      Native.FreeBuffer(cipher_ptr);

      err = Native.ChaCha20Poly1305DecryptAot(
        cipher, cipher.Length, key.Ptr, key.Length,
        associat, associat.Length,
        out IntPtr decipher_ptr, out int decipher_length);

      AssertError(err);
      var decipher = ToBytes(decipher_ptr, decipher_length);
      Native.FreeBuffer(decipher_ptr);

      if (!decipher.SequenceEqual(plain))
        throw new Exception();

      if (i % (rounds / 10) == 0)
        Console.Write(".");
    }

    sw.Stop();

    var t = sw.ElapsedMilliseconds;
    Console.Write($" rounds = {rounds}; t = {t}ms; td = {t / rounds}ms");
    Console.WriteLine();
  }

  public static void TestChaCha20Poly1305BytesStress()
  {
    Console.Write($"{nameof(TestChaCha20Poly1305BytesStress)}: ");

    var rand = Random.Shared;
    var sw = Stopwatch.StartNew();

    var size = rand.Next(
      NetServices.MIN_PW_SIZE, 17);

    var pw = RngBytes(size);

    size = rand.Next(
      NetServices.MIN_SALT_SIZE,
      NetServices.MAX_SALT_SIZE);

    var salt = RngBytes(size);

    using var key = new UsIPtr<byte>(
     Rfc2898DeriveBytes.Pbkdf2(
     password: pw, salt: salt,
     iterations: NetServices.MIN_ITERATION,
     hashAlgorithm: HashAlgorithmName.SHA512,
     outputLength: NetServices.AES_GCM_MAX_KEY_SIZE));

    var tag = new byte[NetServices.AES_GCM_TAG_SIZE];
    var nonce = RngBytes(NetServices.AES_GCM_NONCE_SIZE);
    var plain = RngBytes(NetServices.AES_GCM_MAX_PLAIN_SIZE);

    var associat = Encoding.UTF8.GetBytes("© Michele Natale 2021");

    var err = Native.ChaCha20Poly1305EncryptAot(
    plain, plain.Length, key.Ptr, key.Length,
    associat, associat.Length,
    out IntPtr cipher_ptr, out int cipher_length);

    AssertError(err);
    var cipher = ToBytes(cipher_ptr, cipher_length);
    Native.FreeBuffer(cipher_ptr);

    err = Native.ChaCha20Poly1305DecryptAot(
      cipher, cipher.Length, key.Ptr, key.Length,
      associat, associat.Length,
      out IntPtr decipher_ptr, out int decipher_length);

    AssertError(err);
    var decipher = ToBytes(decipher_ptr, decipher_length);
    Native.FreeBuffer(decipher_ptr);

    sw.Stop();

    if (!decipher.SequenceEqual(plain))
      throw new Exception();

    Console.Write($"t = {sw.ElapsedMilliseconds}ms");
    Console.WriteLine();
  }

  #endregion Native


  #region Managed
  public static void TestChaCha20Poly1305BytesManaged(int rounds)
  {
    Console.Write($"{nameof(TestChaCha20Poly1305BytesManaged)}: ");

    var rand = Random.Shared;
    var sw = Stopwatch.StartNew();
    for (int i = 0; i < rounds; i++)
    {
      var size = rand.Next(
        NetServices.MIN_PW_SIZE, 17);

      var pw = RngBytes(size);

      size = rand.Next(
        NetServices.MIN_SALT_SIZE,
        NetServices.MAX_SALT_SIZE);

      var salt = RngBytes(size);

      using var key = new UsIPtr<byte>(
        Rfc2898DeriveBytes.Pbkdf2(
        password: pw, salt: salt,
        iterations: NetServices.MIN_ITERATION,
        hashAlgorithm: HashAlgorithmName.SHA512,
        outputLength: NetServices.CHACHA_POLY_MAX_KEY_SIZE));

      var tag = new byte[NetServices.CHACHA_POLY_TAG_SIZE];
      var nonce = RngBytes(NetServices.CHACHA_POLY_NONCE_SIZE);

      size = rand.Next(
        NetServices.CHACHA_POLY_MIN_PLAIN_SIZE,
        NetServices.CHACHA_POLY_MAX_PLAIN_SIZE);

      var plain = RngBytes(size);

      var associat = Encoding.UTF8.GetBytes("© Michele Natale 2021");

      var err = CryptoBridge.ChaCha20Poly1305EncryptAotManaged(
              plain, key, associat, out var cipher);
      AssertError(err);

      err = CryptoBridge.ChaCha20Poly1305DecryptAotManaged(
        cipher, key, associat, out var decipher);
      AssertError(err);

      if (!decipher.SequenceEqual(plain))
        throw new Exception();

      if (i % (rounds / 10) == 0)
        Console.Write(".");
    }

    sw.Stop();

    var t = sw.ElapsedMilliseconds;
    Console.Write($" rounds = {rounds}; t = {t}ms; td = {t / rounds}ms");
    Console.WriteLine();
  }

  public static void TestChaCha20Poly1305BytesStressManaged()
  {
    Console.Write($"{nameof(TestChaCha20Poly1305BytesStressManaged)}: ");

    var rand = Random.Shared;
    var sw = Stopwatch.StartNew();

    var size = rand.Next(
      NetServices.MIN_PW_SIZE, 17);

    var pw = RngBytes(size);

    size = rand.Next(
      NetServices.MIN_SALT_SIZE,
      NetServices.MAX_SALT_SIZE);

    var salt = RngBytes(size);

    using var key = new UsIPtr<byte>(
     Rfc2898DeriveBytes.Pbkdf2(
     password: pw, salt: salt,
     iterations: NetServices.MIN_ITERATION,
     hashAlgorithm: HashAlgorithmName.SHA512,
     outputLength: NetServices.AES_GCM_MAX_KEY_SIZE));

    var tag = new byte[NetServices.AES_GCM_TAG_SIZE];
    var nonce = RngBytes(NetServices.AES_GCM_NONCE_SIZE);
    var plain = RngBytes(NetServices.AES_GCM_MAX_PLAIN_SIZE);

    var associat = Encoding.UTF8.GetBytes("© Michele Natale 2021");

    var err = CryptoBridge.ChaCha20Poly1305EncryptAotManaged(
            plain, key, associat, out var cipher);
    AssertError(err);

    err = CryptoBridge.ChaCha20Poly1305DecryptAotManaged(
      cipher, key, associat, out var decipher);
    AssertError(err);

    sw.Stop();

    if (!decipher.SequenceEqual(plain))
      throw new Exception();

    Console.Write($"t = {sw.ElapsedMilliseconds}ms");
    Console.WriteLine();
  }

  #endregion Managed
}

File: C-Abi-Bright-Aot/Tests/CAbiBridge.Aot.Vb.Tests/TestCrypto/TestCryptoChaCha20Poly1305/CryptoChaCha20Poly1305BytesTest.vb
Option Strict On
Option Explicit On

Imports System.Text
Imports michele.natale
Imports michele.natale.Pointers
Imports System.Security.Cryptography

Namespace michele.natale.Tests
  Partial Class CryptoChaCha20Poly1305Test
    Public Shared Sub TestChaCha20Poly1305Bytes(rounds As Int32)
      Console.Write($"{NameOf(TestChaCha20Poly1305Bytes)}: ")

      Dim rand = Random.[Shared]
      Dim sw = Stopwatch.StartNew()
      Dim cipher_ptr As IntPtr = Nothing,
        cipher_length As Int32 = Nothing,
        decipher_ptr As IntPtr = Nothing,
        decipher_length As Int32 = Nothing

      For i As Int32 = 0 To rounds - 1
        Dim size = rand.[Next](NetServices.MIN_PW_SIZE, 17)
        Dim pw = RngBytes(size)
        size = rand.[Next](NetServices.MIN_SALT_SIZE, NetServices.MAX_SALT_SIZE)
        Dim salt = RngBytes(size)

        Using key = New UsIPtr(Of Byte)(
          Rfc2898DeriveBytes.Pbkdf2(
            password:=pw, salt:=salt,
            iterations:=NetServices.MIN_ITERATION,
            hashAlgorithm:=HashAlgorithmName.SHA512,
            outputLength:=NetServices.CHACHA_POLY_MAX_KEY_SIZE))

          Dim nonce = RngBytes(NetServices.CHACHA_POLY_NONCE_SIZE)
          Dim tag = New Byte(NetServices.CHACHA_POLY_TAG_SIZE - 1) {}

          size = rand.[Next](NetServices.CHACHA_POLY_MIN_PLAIN_SIZE,
            NetServices.CHACHA_POLY_MAX_PLAIN_SIZE)

          Dim plain = RngBytes(size)
          Dim associat = Encoding.UTF8.GetBytes("© Michele Natale 2021")

          Dim err = ChaCha20Poly1305EncryptAot(
            plain, plain.Length,
            key.Ptr, key.Length,
            associat, associat.Length,
            cipher_ptr, cipher_length)
          AssertError(err)

          Dim cipher = ToBytes(cipher_ptr, cipher_length)
          FreeBuffer(cipher_ptr)

          err = ChaCha20Poly1305DecryptAot(
            cipher, cipher.Length,
            key.Ptr, key.Length,
            associat, associat.Length,
            decipher_ptr, decipher_length)
          AssertError(err)

          Dim decipher = ToBytes(decipher_ptr, decipher_length)
          FreeBuffer(decipher_ptr)

          If Not decipher.SequenceEqual(plain) Then
            Throw New Exception()
          End If

        End Using

        If i Mod (rounds / 10) = 0 Then
          Console.Write(".")
        End If

      Next

      sw.[Stop]()
      Dim t = sw.ElapsedMilliseconds
      Console.Write($" rounds = {rounds}; t = {t}ms; td = {t / rounds}ms")
      Console.WriteLine()
    End Sub

    Public Shared Sub TestChaCha20Poly1305BytesStress()
      Console.Write($"{NameOf(TestChaCha20Poly1305BytesStress)}: ")

      Dim rand = Random.[Shared]
      Dim sw = Stopwatch.StartNew()

      Dim size = rand.[Next](NetServices.MIN_PW_SIZE, 17)
      Dim pw = RngBytes(size)

      size = rand.[Next](NetServices.MIN_SALT_SIZE,
        NetServices.MAX_SALT_SIZE)
      Dim salt = RngBytes(size)

      Using key = New UsIPtr(Of Byte)(
        Rfc2898DeriveBytes.Pbkdf2(
          password:=pw, salt:=salt,
          iterations:=NetServices.MIN_ITERATION,
          hashAlgorithm:=HashAlgorithmName.SHA512,
          outputLength:=NetServices.CHACHA_POLY_MAX_KEY_SIZE))

        Dim plain = RngBytes(1024 * 1024)
        Dim nonce = RngBytes(NetServices.CHACHA_POLY_NONCE_SIZE)
        Dim tag = New Byte(NetServices.CHACHA_POLY_TAG_SIZE - 1) {}
        Dim associat = Encoding.UTF8.GetBytes("© Michele Natale 2021")
        Dim cipher_ptr As IntPtr = Nothing, cipher_length As Int32 = Nothing

        Dim err = ChaCha20Poly1305EncryptAot(
          plain, plain.Length,
          key.Ptr, key.Length,
          associat, associat.Length,
          cipher_ptr, cipher_length)
        AssertError(err)

        Dim cipher = ToBytes(cipher_ptr, cipher_length)
        FreeBuffer(cipher_ptr)

        Dim decipher_ptr As IntPtr = Nothing, decipher_length As Int32 = Nothing
        err = ChaCha20Poly1305DecryptAot(
          cipher, cipher.Length,
          key.Ptr, key.Length,
          associat, associat.Length,
          decipher_ptr, decipher_length)
        AssertError(err)

        Dim decipher = ToBytes(decipher_ptr, decipher_length)
        FreeBuffer(decipher_ptr)

        sw.[Stop]()

        If Not decipher.SequenceEqual(plain) Then
          Throw New Exception()
        End If

      End Using

      Console.Write($"t = {sw.ElapsedMilliseconds}ms")
      Console.WriteLine()

    End Sub
  End Class
End Namespace
